### PR TITLE
Add Mapper SDK

### PR DESCRIPTION
**Pull request by QiQi-OvO**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it:**
mapper SDK helps developers reduce the workload of developing mapper.

**Which issue(s) this PR fixes:**
Fixes https://github.com/kubeedge/mappers-go/issues/69

**Special notes for your reviewer:**
1. You can find more details in [docs](https://github.com/QiQi-OvO/mappers-go/tree/main/docs).
2. You can find a demo named virtualdevice-sdk in [mappers/virtualdevice-sdk](https://github.com/QiQi-OvO/mappers-go/tree/main/mappers/virtualdevice-sdk)

### CODE REVIEW

**Comment by RyanZhaoXB** [comment]
hello, Qi @QiQi-OvO  . your commit should be provided with sign-off information. you can try to re-commit with this command `git commit -s -m "XXX"` or `git commit -s --amend`  and push your commit again.

File: mapper-sdk-go/internal/application/callback.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+		return common.KindEntityDoesNotExist
+	}
+	return ""
+}
+
+// ReadDeviceData internal callback function
+func ReadDeviceData(deviceID string, propertyName string, dic *di.Container) (response string, kind common.ErrKind) {
+	deviceInstance := instancepool.DeviceInstancesNameFrom(dic.Get)
+	if _, ok := deviceInstance[deviceID]; !ok {
+		kind := common.KindEntityDoesNotExist
```
Comment (RyanZhaoXB) [inline mapper-sdk-go/internal/application/callback.go:240]: kind已经作为显式命名的返回值，这里是不是可以不用重新声明一个kind，或者直接return common.KindEntityDoesNotExist

File: mapper-sdk-go/internal/application/callback.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	}
+	if index == -1 {
+		kind := common.KindEntityDoesNotExist
+		return "", kind
+	}
+	mapMutex := instancepool.DeviceLockNameFrom(dic.Get)
+	protocolDriver := instancepool.ProtocolDriverNameFrom(dic.Get)
+	response, err := controller.GetDeviceData(deviceID, deviceInstance[deviceID].Twins[index], protocolDriver, mapMutex[deviceID], dic)
+	if err != nil {
+		kind = common.KindServerError
```
Comment (RyanZhaoXB) [inline mapper-sdk-go/internal/application/callback.go:258]: 有error可以报个log，279行同

**Review by RyanZhaoXB — CHANGES_REQUESTED** [review]

**Review by RyanZhaoXB — APPROVED** [review]
lgtm

**Comment by RyanZhaoXB** [comment]
@fisherxu please help to review this pr. thanks.

**Comment by RyanZhaoXB** [comment]
@chendave @fisherxu please help to review this pr. thanks.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,11 @@
 # mappers-go
 KubeEdge Device Mappers written in go, it works for KubeEdge 1.4 version and later.
+
+## Construction method
+### 1 mappers
+
+
+### 2 mapper-go-sdk
+
+SDK provide RESTful and you can connect your device fastly.
```
Comment (fisherxu) [inline README.md:10]: What's meaning of `SDK provide RESTful`? It's better to introduce how the the mapper SDK work here.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,11 @@
 # mappers-go
 KubeEdge Device Mappers written in go, it works for KubeEdge 1.4 version and later.
+
+## Construction method
+### 1 mappers
```
Comment (fisherxu) [inline README.md:5]: What's meaning of `Construction method` here? Seems there's no content under mappers title.

It's better not to use the number like `1` in markdown

File: docs/MapperSDKDesign.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
+# Mapper-sdk-go Design
```
Comment (fisherxu) [inline docs/MapperSDKDesign.md:1]: The file name should be `mapper-sdk-design.md`

File: docs/UserGuideofMapperSDK.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# User Guide of Customized Mapper SDK in KubeEdge
```
Comment (fisherxu) [inline docs/UserGuideofMapperSDK.md:1]: ditto

File: _template/mapper-sdk/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,6 @@
+FROM alpine:3.14
+RUN mkdir -p kubeedge
+COPY ./bin/Template kubeedge/bin/
```
Comment (fisherxu) [inline _template/mapper-sdk/Dockerfile:3]: Should include `make` in all dockerfile, users can make the image from code directly

Comment (QiQi-OvO) [reply _template/mapper-sdk/Dockerfile:3]: Thank you for your valuable advice, I am trying to correct other mistakes : ).  
For this Dockerfile,the corresponding operation has been given in [README](https://github.com/QiQi-OvO/mappers-go/blob/main/_template/mapper-sdk/README.md), ```make mapper Template package``` can build the code and copy the compiled file to the image. Mapper and mapper-sdk-go both follow this rule, so I think it is reasonable to follow the previous ***make rules*** and ***Dockerfile***.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,15 @@
...
+are provided, the basic ```mappers``` and ```mapper-sdk-go```.
+### mappers
+Mappers implement modbus, bluetooth, onvif and opcua. If you want to connect devices to kubeedge of these protocols, you can find
+them directly in [mappers](./mappers).
+### mapper-go-sdk
+Mapper-sdk-go is written based on the basic mapper, which provides a convenient RESTful interface. Users can quickly 
+get device data and device status through these interfaces.
+Mapper-sdk-go supports connect customized devices to kubeedge fastly. If you want to access a device with a new protocol, 
+mapper-sdk-go is a good choice.
+You can get more information and details in [mapper-sdk-go](./mapper-sdk-go/)
```
Comment (fisherxu) [inline README.md:15]: From the description, I'm not quite understand, users can use mapper-go-sdk diectly without any development?

From the name `mapper-go-sdk`, it's a lib, users can import it to implement the mapper.

`Mapper-sdk-go is written based on the basic mapper`, mapper-sdk-go is based on mapper? I think mapper-sdk-go should provide some basic libs or framework to build a new mapper. Users can build a new mapper easily with mapper-sdk-go, only need to implement the customized protocol driver.

Need to update the descriptions above

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**